Grain merging considering misorientation angle

PengWei97 opened the pull request:
This pull request (https://github.com/idaholab/moose/pull/23757) introduces a new feature to GrainTracker that enables merging of grains based on misorientation.

## Reason
When two grains share the same crystallographic orientation or have a misorientation angle below a very small threshold, they should merge into a single grain.

## Design
First, adjacent_id is added to FeatureFloodCount to store the adjacent Grain ID. Second, ...

## Impact
The addition of FeatureFloodCount and GrainTracker, as well as the creation of a new MOOS object (Object X), help to enhance the grain merging functionality. This capability is part of an effort to consider grain growth based on EBSD microstructures.


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

PengWei97 committed:
To output the feature's topological information, modify
FeatureFloodCount.

PengWei97 committed:
Merge branch 'idaholab:next' into next

PengWei97 committed:
Merge branch 'idaholab:next' into next

PengWei97 committed:
Remove redundant comments

PengWei97 committed:
Merge branch 'next' of github.com:PengWei97/moose into next

PengWei97 committed:
Get adjacent grain vector in GrainTracker

PengWei97 committed:
Add interface for grain merging in GrainTracker

PengWei97 committed:
create GrainTrackerMerge based on GrainTracker

PengWei97 committed:
Added grain merging functionality in GrainTracker

PengWei97 committed:
Add a test input file for grain merging

permcody commented on modules/phase_field/include/postprocessors/GrainTracker.h:
```diff
@@ -111,6 +117,15 @@ class GrainTracker : public FeatureFloodCount, public GrainTrackerInterface
+  virtual void mergeGrainsBasedMisorientation();
+
+  const EulerAngleProvider & _euler;
+  misoriAngle_isTwining _s_misoriTwin;  
```
This doesn't follow our naming conventions. Normally for classes and methods we use CamelCase exclusively and for variables we'll use underscores to separate words. The type here is a combination of those two conventions. Let's switch it to CamelCase.

PengWei97 replied:
Based on your suggestion as well as [Naming](https://mooseframework.inl.gov/workshop/index.html#/14/4), I normalized a few names:  misoriAngle_isTwining  to MisorientationAngleData; _s_misoriTwin to _s_misorientation_angle; quatReal to QuatReal; _merge_grains_basedMisorAngle to _merge_grains_based_misorientaion;

permcody commented on modules/phase_field/include/postprocessors/GrainTracker.h:
```diff
@@ -111,6 +117,15 @@ class GrainTracker : public FeatureFloodCount, public GrainTrackerInterface
+  misoriAngle_isTwining _s_misoriTwin;  
+  
+  // True if two grains are determined to perform a merge operation
+  bool _merge_grains_basedMisorAngle;  
```
Same here - all underscores, or all CamelCase.

PengWei97 replied:
areadly edited

PengWei97 replied:
already edited

permcody commented on modules/phase_field/include/postprocessors/GrainTrackerMerge.h:
```diff
@@ -0,0 +1,13 @@
+
+#include "GrainTracker.h"
+
+class GrainTrackerMerge : public GrainTracker
```
I spoke with one of our phase field experts and we believe that all of your new logic should just go in this class to leave the existing classes alone to the extent that's possible. You will have to add your new adjacency members to the `FeatureData` structure, and you may also need a new callback, but most of the rest of this can go in this class to separate it out. 

This will also allow you to have more of a "development" object where you can put your work without affecting the base classes. It'll be easier for your to make changes and get them merged.

PengWei97 replied:
Yeah, I also agree with that the new logic of determining whether the grains are merged should be included in GrainTrackerMerge.

> You will have to add your new adjacency members to the FeatureData structure, and you may also need a new callback,

I am not quit sure what yuo mean by "a new callback". As you can see, I created the `_adjacent_id` direcly in the `FeatureData` of the `FeatureFloodCount` class. This has not been moved to the new GrainTrackMerge , which I think it would be reasonable now for the following reasons:

1. Once materials or vectorpostprocessors objects may also need to access this part  of the data, such as a vectorpostprocessors class that can obtain the number of adjacent grains of each grain to verify the class  von Neumann-Mullins theory in grain growth.
2. If this part is moved to GrainTrackerMerge, the specific operation is a bit challenging for me.

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -78,6 +81,8 @@ GrainTracker::validParams()
   : FeatureFloodCount(parameters),
     GrainTrackerInterface(),
+    _merge_grains_basedMisorAngle(getParam<bool>("merge_grains_basedMisorAngle")),
+    _euler(getUserObject<EulerAngleProvider>("euler_angle_provider")),
```
This is one of the big reasons we don't want to modify this class. The GrainTracker works for toy Voronoi problems where we don't care about orientation. We don't want to enforce that every use of this class now have to consider Euler Angles. This requirement should be added to your new GrainTrackerMerge class.

PengWei97 replied:
Yeah, I have fallen back to a version that didn't add this part of the functionality and tried implementing it in GrainTrackerMerge;

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -866,6 +914,33 @@ GrainTracker::trackGrains()
 }
 
+void 
+GrainTracker::createAdjacentIDVector() // by weipeng
```
This is good but might be good to move to your new class. Again, most of the base class usage doesn't require this and this might start to be a time sync for systems with very large amounts of grains since the "haloIntersect" method can be expensive.

PengWei97 replied:
OK， I took your advice and moved createAdjacentIDVector into GrainTrackerMerge.

permcody commented on modules/phase_field/include/utils/MisorientationAngleCalculator.h:
```diff
@@ -0,0 +1,36 @@
+ * This is an orientation difference calculator, inspired by MTEX.
+ * However, it is an incomplete version and needs further optimization TODO.
+ */
+class MisorientationAngleCalculator
```
You might consider making this a namespace instead. There are no instance variables nor instance methods. It's just a list of static methods.

PengWei97 replied:
already edited

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+  {
+    value_acos = dotQuaternion(mori_q1q2, q3_twin[i], qcs, qcs);
+    if (value_acos <= 1.0 && value_acos >= -1.0)
+      misor_twinning = (Real)(2.0*std::acos(value_acos))/degree;
```
```suggestion
      misor_twinning = Real((2.0*std::acos(value_acos))/degree);
```

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+    if (value_acos <= 1.0 && value_acos >= -1.0)
+      misor_twinning = (Real)(2.0*std::acos(value_acos))/degree;
+
+    s.isTwinning = (bool)(misor_twinning < tolerance_mis); // Judging whether it is a twin boundary
```
```suggestion
    s.isTwinning = misor_twinning < tolerance_mis; // Judging whether it is a twin boundary
```

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+misoriAngle_isTwining
+MisorientationAngleCalculator::calculateMisorientaion(EulerAngles & Euler1, EulerAngles & Euler2, misoriAngle_isTwining & s, const std::string & CrystalType)
+{
+  const Real degree = 1.7453e-02;
```
Oh, is this just a conversion from Radians to degrees?

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+  // calculate misorientation angle
+  value_acos = dotQuaternion(q1, q2, qcs, qss);
+  if (value_acos <= 1.0 && value_acos >= -1.0)
+    s.misor = (Real)(2.0*std::acos(value_acos))/degree;
```
```suggestion
    s.misor = (2.0*std::acos(value_acos))/degree;
```

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+    // Determine which type of twin boundary 0 ~ TT1 (tensile twins), 1 ~ CT1 (compression twins)
+    if (s.isTwinning) 
+    {
+      s.twinType = "twin_type" + std::to_string(i); 
```
You should consider using `enum`s instead of string comparisons for this and all of the logic in `getKeyQuat`. Just a suggestion

PengWei97 replied:
Create enumerated types TwinType, QuatType and CrystalType in MisorientationAngleData

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+    }; // 24 symmetric for fcc
+
+  std::vector<quatReal> q(q_num.size());
+  for (unsigned int i = 0; i < q_num.size(); ++i)
```
```suggestion
  for (const auto i : index_range(q_num)
```

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+  a1 = q1.w(); b1 = -q1.x(); c1 = -q1.y(); d1 = -q1.z();
+  a2 = q2.w(); b2 = q2.x(); c2 = q2.y(); d2 = q2.z();
+
+  // standart algorithm
```
```suggestion
  // standard algorithm
```

permcody commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+{
+  std::vector<Real> d_vec(rot2.size());
+
+  for (unsigned int i = 0; i < rot2.size(); ++i)
```
```suggestion
  for (const auto i : index_range(rot2))
```

permcody reviewed (changes requested):
This is a good start, but overall I'd like you to move most of your new logic into your new class as you continue to develop. I think it makes sense to extend the base classes since your new capability requires Euler angles to work properly. That's not something the base classes require now. 

Also for your testing, you'll need to put in a "tests" file to actually run your new test. Look at similar examples in other test folders to see how you set these up. We'll also eventually need documentation (GrainTrackerMerge.md) files. 

Keep going!

PengWei97 committed:
add merge function in GrainTrackerMerge

PengWei97 committed:
Create GrainTrackerMerge.md

PengWei97 commented on modules/phase_field/include/postprocessors/FeatureFloodCount.h:
```diff
@@ -83,6 +83,14 @@ class FeatureFloodCount : public GeneralPostprocessor,
   virtual unsigned int getFeatureVar(unsigned int feature_id) const;
 
+  /// Returns the feature ID representing the passed in feature (Grain ID)
+  virtual unsigned int getFeatureID(unsigned int feature_id) const;
```
You are right, I read this function carefully later and found that it is not needed, now I remove it;

PengWei97 committed:
Eliminate warnings due to incorrect order

PengWei97 committed:
Merge branch 'grainMergeMisorThreshold' of github.com:PengWei97/moose into grainMergeMisorThreshold

PengWei97 commented on modules/phase_field/include/postprocessors/FeatureFloodCount.h:
```diff
@@ -170,7 +179,8 @@ class FeatureFloodCount : public GeneralPostprocessor,
         _status(status),
-        _boundary_intersection(BoundaryIntersection::NONE)
+        _boundary_intersection(BoundaryIntersection::NONE),
+        _adjacent_id(adjacent_id)
```
OK，I've changed the order of them and it runs without warnings.

PengWei97 committed:
Canonicalized naming

PengWei97 committed:
Normalization named _merge_grains_basedMisorAngle

PengWei97 commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -839,6 +841,63 @@ FeatureFloodCount::getFeatureVar(unsigned int feature_id) const
 }
 
+unsigned int
+FeatureFloodCount::getFeatureID(unsigned int feature_id) const
```
Yeah, I have been removed the function getFeatureID;

PengWei97 committed:
modify the function name of getAdjacentGrainNum

PengWei97 committed:
move createAdjacentIDVector into GrainTrackerMerge from GrainTracker

PengWei97 commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,200 @@
+MisorientationAngleCalculator::calculateMisorientaion(EulerAngles & Euler1, EulerAngles & Euler2, misoriAngle_isTwining & s, const std::string & CrystalType)
+{
+  const Real degree = 1.7453e-02;
+  Real tolerance_mis = 3.90;
```
already edited

PengWei97 committed:
Convert MisorientationAngleCalculators into a namespace

PengWei97 committed:
Change _twin_type to enumeration type

PengWei97 committed:
Create enumeration types QuatType and CrystalType in MisorientationAngleData

PengWei97 committed:
Some small problems were modified, according to cody's suggestion

PengWei97 commented:
Hi @permcody,

According to your suggestion, I modified an imperfect place today. The following thing is what I need to do next,

- [ ] Is it reasonable to put _adjacent_id into FeatureFloodData?
- [ ] Use debugging method to solve some bugs mentioned in #22853.
- [ ] Testing with EBSD initial structure (optimistic estimate to be completed next week)

Also, please forgive me for not being very familiar with such a development process in Github, I will have a morning implementation over the weekend to get familiar with it.

Have a great weekend！

wei

dschwen commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -839,6 +841,63 @@ FeatureFloodCount::getFeatureVar(unsigned int feature_id) const
+std::vector<unsigned int>
+FeatureFloodCount::getAdjacentID(unsigned int feature_id) const
+{ 
+  std::vector<unsigned int> invalid_id_vector(10,invalid_id);
```
why 10?

PengWei97 replied:
Oh, this is a randomly selected number. At that time, it was considered that in 2D, the data of adjacent grains of each grain rarely exceeds 10, so 10 was selected. And, the creation of this function refers to getFeatureVar.

Looking back now, it is more sensible to get set to 1

dschwen commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,212 @@
+    return d;
+  }
+
+  QuatReal itimesQuaternion(const QuatReal & q1, const QuatReal & q2)
```
sure that is doesn't already exist in libEigen?

PengWei97 replied:
Yes, as I said before, I didn't make full use of the existing c++ library (libEigen), and I will optimize this part later.

dschwen commented on modules/phase_field/src/utils/MisorientationAngleCalculator.C:
```diff
@@ -0,0 +1,212 @@
+
+  Real dotOuterQuaternion(const QuatReal & rot1, const std::vector<QuatReal> & rot2)
+  {
+    std::vector<Real> d_vec(rot2.size());
```
creating a local vector is very slow, you'd be better off to compute the largest element on the fly.

PengWei97 replied:
Okay, you mean to use this calculation method?

```c++
    Real d = 0;
    Real temp = 0;
    for (const auto i : index_range(rot2))
    {
      temp = std::abs(rot1.w()*rot2[i].w() + rot1.x()*rot2[i].x() + rot1.y()*rot2[i].y() + rot1.z()*rot2[i].z()); // rot1 * rot2'
      if (temp > d)
        d = temp;
    }

    return d;
```

PengWei97 committed:
Eliminate warnings in MisorientationAngleCalculator

PengWei97 committed:
Solved 2 two bugs mentioned in #22853

PengWei97 commented on modules/phase_field/include/utils/MisorientationAngleCalculator.h:
```diff
@@ -0,0 +1,45 @@
+  QuatReal itimesQuaternion(const QuatReal & q1, const QuatReal & q2);
+
+  // function 3.3: computes outer inner product between two quaternions
+  Real dotOuterQuaternion(const QuatReal & rot1, const std::vector<QuatReal> & rot2);
```
rot2 is actually used to store qcs, which represent crystallographic symmetry matrices. For example, in HCP there are 24 symmetry matrices, each corresponding to a different quaternion. So rot2 stores a sequence of quaternions, each representing a rotation matrix. 

 By the way, the code for the `MisorientationAngleCalculator` is a direct translation from Matlab to C++ of the corresponding part in MTEX. Although I realize the code here is not clean enough, it currently meets the basic requirements, which include:
1. Calculating the misorientaion angle between two sets of inpute Euler angles;
2. Identifying two types  of twins in HCP;

Regarding the optimization of the MisorientationAngleCalculator, I plan to work on it after completing the main task of GrainTrackerMerge.

wei

PengWei97 committed:
Fixed extract_adjacent_id bug in GrainTrackerMerge

PengWei97 committed:
add a test using EBSD for GrainTrackerMerge

PengWei97 committed:
Add the information output by test_GrainTrackerMerge_usingEBSD to the terminal

PengWei97 commented:
Hi @permcody @dschwen,

On the basis that I think the test based on the VT microstructure is completed, I use the EBSD microstructure (pure alpha-titanium) as the initial input structure to test `GrainTrackerMerge`. The test file name is [test_GrainTrackerMerge_usingEBSD.i](https://github.com/idaholab/moose/blob/fc371e79f5dd960819b262f2118fc85a844aae20/modules/combined/test/tests/grain_merge_based_misorientation/test_GrainTrackerMerge_usingEBSD.i).

After performing grain merging (gr17 and gr41) based on misorientation using GrainTrackerMerge, it was discovered that Grain `#17 intersects Grain #63 (variable index: 4) needs to be remapped`. However, the figure below shows that `Remapping grain #17` was not executed."

![image](https://user-images.githubusercontent.com/56183775/226166378-c3b28f6d-f44c-4f01-b01f-7f7150f5d637.png)

In addition, some excerpts of the output to the terminal results below,
```bash
...
Time Step 2, time = 0.21, dt = 0.11
...
Grains active index 9: 2 -> 2

[32mMarking Grain 0 as INACTIVE (variable index: 3)
...

Time Step 3, time = 0.331, dt = 0.121
...
[39m[33mGrain #17 and Grain #41 was merged (misor: 0.601591).
...
[39m[33mSplit Grain (#17) detected on unmatched OPs (2, 4) attempting to remap to 2.
...
[39m[33mGrain #17 intersects Grain #63 (variable index: 4)
[39m[32m- Depth 0: Remapping grain #17 from variable index 4 to 9 whose closest grain (#176) is at a distance of 63.189

[39mFinished inside of GrainTracker
...

Time Step 4, time = 0.4641, dt = 0.1331
...
[39m[32mMarking Grain 63 as INACTIVE (variable index: 4)
...    
```

In general, I don't quite understand why after remapping is detected, the operation from eta4 to eta9 is not performed on grain 17, but the grain 41 is invalidated in Step 4.

PengWei97 committed:
delet 02_ebsd_error.md

permcody commented:
Without running the example, we might just have some edge case due to multiple logic conditions all executing at once here. From your output, the first thing that happens is that #17 and #41 get merged by your new code:

`Grain #17 and Grain #41 was merged (misor: 0.601591).`

However, because of this merge, the "split grain" logic kicks in to keep those two pieces mapped to the same OP. At this stage they aren't physically connected.

`Split Grain (#17) detected on unmatched OPs (2, 4) attempting to remap to 2.`

I know these are all my diagnostic messages, but "attempting to remap" doesn't exactly leave things in a settled state. I'm fairly certain if there had been a failure it would have been reported next. So I'm assuming that #17 and #41 are both on OP 2 now...

So it appears that the problem might be that we just aren't updating the OP number in the data structure? If my assumption above is correct, #17 should be on OP 2 now, but the next message indicates otherwise:

`Grain #17 intersects Grain #63 (variable index: 4)
Depth 0: Remapping grain #17 from variable index 4 to 9 whose closest grain (#176) is at a distance of 63.189`

This could explain why you aren't seeing any remapping. If the data structure just has the wrong information all of the logic to move those values from 4 to 9 would still happen, except there wouldn't be any values to move there (from OP 4) because that grain is on OP 2.

TODO: Can you look into this problem and see if this is the case? If you look at the existing remapping logic, there is a communication step that occurs after the remapping to make sure that all the processor's view of the order parameter mappings are consistent. It's just the communication of a simple map of what just got moved. You may need to do that in your new "misorientation merge" logic. 

Report back if you have any questions.

PengWei97 commented:
Hi @permcody,

According to your suggestion, I tried it for a day today, but unfortunately this problem has not been effectively resolved.

In order to confirm that the eta of the original Grain #17 was successfully set to eta2 from eta4, I created `remapGrainsMerge()` in `GrainTracker` based on the function `remapGrains`, the only difference from `remapGrains` is that there is no remapping when the grain ID is different but the eta is the same.

First, I created the first test and set `remap_grains = false`, the results are shown below,
![image](https://user-images.githubusercontent.com/56183775/226623842-2b121dec-9756-4791-97b1-b51e45dce818.png)
The results show that in the process from Step 2 to Step 3, the original eta of the Grain #17 area is remapped from eta4 to eta2, and the Grain ID of the original Grain #41 area is set to Grain #17. So I think `mergeGrainsBasedMisorientation()` and `remapGrainsMerge()` successfully completed the operation of merging Grains #17 and #41 based on misorientation. Also, since `remap_grains = false`, it is reasonable for Grain #2 not to perform remapping. Things seem to be optimistic so far.

In the second step, I created a second test and set `remap_grains = true`, the problem was not solved, in detail, even though `remapGrains()` detected

```bash
Grain #4 intersects Grain #17 (variable index: 2)
- Depth 0: Remapping grain #4 from variable index 2 to 8 whose closest grain (#54) is at a distance of 65.0187
```

But in the result, Grain #4 did not remap to eta8 from eta2, and finally merged into Grain #17 in the fourth step;

Afterwards, I thought about what we needed to do in detail, and drew a flow chart like this:
![image](https://user-images.githubusercontent.com/56183775/226627790-e31d598c-0d7c-41d4-a504-e849b6f909df.png)

Executing GrainTracker in Step 3 should have three sub-steps:
- Step 3.1 In mergeGrainsBasedMisorientation(), based on misorientation, set `Grain_j._id = Grain_i._id`;
- Step 3.2 In remapGrainsMerge(), when `grain1._id = grain2._id`, remap grain2._var_index = grain1._var_index;
- Step 3.3 execute `remapGrains()`;

Obviously, the result shows that Step 3.3 is not successful, even if the detection requires remapping, it shows the error directed by the arrow in the above figure.

In general, the above results and analysis prove that the operation of merging grains based on misorientation is successful, and the related eta is also remapped. The only problem is that when it is detected that the grain needs to be remapped, it is not successfully executed in `remapGrains()`.   

For your convenience pointing out my possible silly mistakes, the relevant documentation is also provided:

- [test_GrainTrackerMerge_usingEBSD.txt](https://github.com/idaholab/moose/files/11029200/test_GrainTrackerMerge_usingEBSD.txt)
- [case2_EBSD_s1_v1_out.txt](https://github.com/idaholab/moose/files/11029206/case2_EBSD_s1_v1_out.txt)
- [case2_EBSD_s1_v2_out.txt](https://github.com/idaholab/moose/files/11029211/case2_EBSD_s1_v2_out.txt)

I would sincerely appreciate any guidance or answers you could provide me regarding this matter.

wei

PengWei97 committed:
create remapGrainsMerge()

permcody commented:
This is a great test case. Thanks for putting together this detailed analysis. The good news is that the remap routine should be able to handle this complicated case. This is precisely why it is setup recursively. There are many cases where one remap may trigger additional remaps because the order parameters are being reused heavily in a localized area.

I’ll try to find some time to take a closer look at your changes. I think you are close. I wonder if something in the data structure just isn’t in the state you think it’s in based on your flow chart. If you have some time. I might suggest you put in a loop to print out the whole grain structure. I’ve overloaded the stream output operator so you can just put code like “std::cout << grain”, and it’ll dump the ID, current order parameter, and a bunch of other information about halos and which processors own a piece of that grain. You could print out that data structure in a simple loop wherever you want to inspect information.

I’ll take a closer look and let you know if I have additional suggestions when I’m in the office.  Best Regards

PengWei97 commented:
> I’ll try to find some time to take a closer look at your changes. 
Thanks a lot for your help.  I will wrap the code into GrainTrackerMerge as soon as possible, and this way it's easier for you to check for changes.

> I wonder if something in the data structure just isn’t in the state you think it’s in based on your flow chart.

I'm not sure what flow chart you are referring to this it?
![image](https://user-images.githubusercontent.com/56183775/226781116-1af8efa9-8bb6-4d97-9d1c-8e611e9b4425.png)


> I might suggest you put in a loop to print out the whole grain structure. 

Oh, that seems like a good suggestion. I will try to adopt it.

Today I will try to encapsulate the relevant code into GrainTrackerMerge, good luck. I will also regularly communicate with you about the progress and problems.

Best Regards
Wei

PengWei97 committed:
rebuild the code of re-merging grain function

PengWei97 committed:
add a test for re-marging grains using EBSD data

PengWei97 committed:
modify test_GrainTrackerMerge_usingEBSD

PengWei97 committed:
report a error when using recover function in test_GrainTrackerMerge_usingEBSD.i

PengWei97 committed:
Added option to adopt Distributed Mesh in EBSDMeshGenerator

PengWei97 committed:
Fix a bug in EBSDReader

PengWei97 commented:
Hi @permcody,

Today's luck seems to be exceptionally good. Today I completed four small goals in my model building:

First,[ the code of re-merging grains function](https://github.com/idaholab/moose/pull/23757/commits/aceb2c502852be51c775f3423a4639063dc64f99) in `GrainTracker` AND `GrainTrackerMerge` was rebuilt, specifically I added two hooks in `GrainTracker`. One is located in `trackerGrain()`, named `mergeGrainsBasedMisorientation()`, which is used to detect which adjacent grains need to be merged, when the misorientation angle is lower than a certain threshold; the other is located before the execution of remapGrains(), named `remapGrainsMerge()` is used to remap the grains that need to be merged to the same variable index. Finally, for the specific implementation of these two functions, I overloaded them in `GrainTrackerMerge`;

Next, an example of using EBSD data as the initial microstructure is used to verify the effectiveness of the added function, named [test_GrainTrackerMerge_usingEBSD.i](https://github.com/idaholab/moose/pull/23757/commits/05f5887ce000d9b04aa8a0041a5e21483595dae0) is added to the `combined\examples` directory, and the output is as follows:
![image](https://user-images.githubusercontent.com/56183775/226938537-cad9cc0f-525b-4061-a727-0774392f5a03.png)

Second, in order to facilitate the effective operation of large-scale calculations in the future, the recover function has also been verified. However, the above calculation example shows that there is an error in the code of `EBSDReader`, which is caused by not calling `readFile()` after recover, so[ I deleted the following code](https://github.com/idaholab/moose/pull/23757/commits/196e16fb14fc36ccaeb0530570e72404a12b56e4) in `EBSDReader`;

Third, in order to use Distributed Mesh in the above example, I found that `EBSDReader` and its related codes have no related functions, so I [modified and added](ttps://github.com/idaholab/moose/pull/23757/commits/0c9e159e3cece95ae6558541d79798887aab3939) `use_distributed_mesh` option in `MooseMesh.C`;

Fourth, in order to consider the grain growth driven by stored energy, passing in GNDs in EBSD data seems to be a good strategy. Also based on the above calculation example, set `custom_columns = 1` in `UserObjects\ebsd_reader\`, but there is a data type mismatch error in the following code in `EBSDReader`,
```c++
if (field_name == "CUSTOM" + Moose::stringify(i))
```
[So type casts are performed and it worked. ](https://github.com/idaholab/moose/pull/23757/commits/c8511ede7bd27b9535063ba98fe728df15b204bb)

In addition, in order to accomplish the purpose mentioned above, I also created three Moose Objects `ACSEDGPolyEBSD PolycrystalStoredEnergyEBSDAction DeformedGrainEBSDMaterial` in my own APP-yinglong. If you think this is meaningful, I am also very happy to merge it into Moose.

By the way, there may be a better way around `EBSDReader` for the last three modifications, but I didn't find it. Finally, any comments on improvements to this modification would be greatly appreciated.

Best Regards
Wei

roystgnr commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -75,6 +75,11 @@ MooseMesh::validParams()
                              "REPLICATED: Always use libMesh::ReplicatedMesh "
                              "DISTRIBUTED: Always use libMesh::DistributedMesh");
 
+  params.addParam<bool>(
+      "use_distributed_mesh",
+      true,
+      "If use_distributed_mesh=true, distributed mesh is adopted in calculation.");
```
I'm confused.  How is this intended to differ from the existing `parallel_type=distributed` option?  Where is the new `_use_distributed_mesh` member here used?

PengWei97 replied:
Oh, maybe I made a stupid mistake (in discussion of #22135), in fact, this addition in unnecessary through adding `parallel_type=distributed` in `[Mesh]`. Now I delete this part.

Thank you very much for your comments.

PengWei97 committed:
delet the option `use_distributed_mesh = true`

PengWei97 committed:
modify MisorientationAngleCalculator for identify twins

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -359,6 +361,9 @@ GrainTracker::finalize()
   broadcastAndUpdateGrainData();
 
+  if (_merge_grains_based_misorientaion && _t_step > 2)
+    remapGrainsMerge();
```
The GrainTracker has a feature where we start tracking at a user-defined step (see the top of this method). I don't think it's often used, but you may want to call your method based on an offset of `_tracking_step`. Otherwise the use of that parameter may cause some oddities with the remaining logic in this method. 

Also if I may, I would suggest that we rename this method: "merge" and "remap" are both verbs so it sounds like we are remapping and merging. Some suggestions: "remapMisorientedGrains", "remapMergableGrains", "pairTwinGrains", etc. I believe the idea here is that you are trying to keep "twins" together so the method name should clearly indicate that.

PengWei97 replied:
Ok, I have changed `_t_step > 2` to `_t_step > _tracking_step`, and change remapGrainsMerge to remapMisorientedGrains.

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -802,7 +813,7 @@ GrainTracker::trackGrains()
         if (grain._status == Status::CLEAR)
         {
-          auto new_index = getNextUniqueID();
+          auto new_index = 0; //getNextUniqueID();
```
Probably a mistake, this needs to be changed back for nucleation test cases.

PengWei97 replied:
Oh yes. I accidentally uploaded this.

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -816,6 +827,10 @@ GrainTracker::trackGrains()
     }
 
+    // The reason for _t_step > 2 is to allow reasonable refinement of the mesh at grain boundaries.
+    if (_merge_grains_based_misorientaion && _t_step > 2)
```
Similar comment as above. You'll want to change _t_step to just being an offset from the `_tracking_step` instead.

PengWei97 replied:
Yes, the same modification as above

permcody commented on modules/phase_field/src/postprocessors/GrainTracker.C:
```diff
@@ -1168,6 +1183,11 @@ GrainTracker::remapGrains()
+void
+GrainTracker::remapGrainsMerge()
+{
+  _console << "This function needs to be defined in the derived class\n" << std::endl;
```
You probably don't want to print this message out here. This won't make any sense for existing use of the GrainTracker class. Just leave the method empty.

PengWei97 replied:
Ok, redundant code has been removed.

permcody commented on modules/phase_field/src/postprocessors/GrainTrackerMerge.C:
```diff
@@ -0,0 +1,164 @@
+    _nl.solutionOld().close();
+    _nl.solutionOlder().close();
+
+    _fe_problem.getNonlinearSystemBase().system().update();
```
This code is all copied from GrainTracker. Let's try to reuse that code even if it means moving it to a new method in GrainTracker so you can call it from GrainTracker and this class.

PengWei97 replied:
I determined that these are redundant here by deleting these three codes, and deleted them.

permcody commented on modules/phase_field/src/postprocessors/GrainTrackerMerge.C:
```diff
@@ -0,0 +1,164 @@
+
+    _fe_problem.getNonlinearSystemBase().system().update();
+
+    _console << COLOR_RED << "Swaps complete in remapGrainsMerge" << std::endl;
```
COLOR_RED? You probably should avoid red for "normal" conditions.

PengWei97 replied:
Ok, I changed COLOR_RED to COLOR_GREEN.

permcody commented:
Glad you are making additional progress on this:

1) I suggest that you address the precheck target issues so we can get the tests running on your PR so you'll know if you are close to being able to merge this. Right now you are failing due to whitespace issues. There are instructions for what to do right in the output of the precheck step. If you have questions about how to resolve those issues, just ask. 
2) I'm looking at your images above - they look good. I might look into the string issue you encountered as I don't understand why that change is necessary.
3) On Recover - note that most methods do get rerun on recover, including methods like "InitialSetup" which can be used to do any additional setup you need to perform on continuing your run. Notably as you discovered though, none of the Mesh related objects get rerun. This is because you don't want to start with a new mesh. That's some of the stateful data that gets saved to disk and then reloaded for you. You'll have your mesh in whatever state it was in when you simulation terminated along with all of the solution fields still on it. If you are using EBSDMesh, there's no need to trigger that file read again at least for the purposes of the mesh. However, it should be possible to reread the file for other purposes. I would be surprised if we hadn't already considered that. I would focus on this step towards the end though. Let's get everything else working first.

PengWei97 committed:
code modification based on permbody's

PengWei97 commented:
Hi @permcody,

I totally agree with your suggestion. For the ability to perform grain merging operations based on neighboring grains due to misorientation below a certain threshold, I consider it done.

> 1 - so we can get the tests running on your PR so you'll know if you are close to being able to merge this. There are instructions for what to do right in the output of the precheck step.

I don't know what I need to do in this test session, where is the instruction you mentioned?

> 2 - I might look into the string issue you encountered as I don't understand why that change is necessary.

Ok, very much looking forward to your suggestions and inspections.

> 3 - I would focus on this step towards the end though. Let's get everything else working first.

Yes, I think our first priority is to add the aforementioned merging functionality to GrainTracker. Then consider the recover environment.

Finally, I am very sorry that it took me so long to reply to you. Because I am having a headache for another modeling thing, that is, considering GB anisotropy under the GrainTracker algorithm, based on `GBAnisotropy`, the function addition of this piece will be explained in detail in Discussions later, and I look forward to your guidance.

wei

PengWei97 commented:
Hi @permcody,

I'm pleased to inform you that I've successfully resolved the challenge of developing GB anisotropy code based on misorientation, which had been a major hurdle for me. 

Moving forward, I would like to focus on the work of Grain Merging based on misorientation and would appreciate your valuable feedback on it.

wei

PengWei97 committed:
Merge branch 'idaholab:next' into grainMergeMisorThreshold
